habitat/core/dataset.py

```diff
@@ -324,29 +342,48 @@ def __next__(self):
 
         :return: next episode.
         """
+        self._switch_scene_if()
 
         next_episode = next(self._iterator, None)
         if next_episode is None:
             if not self.cycle:
                 raise StopIteration
+
             self._iterator = iter(self.episodes)
+
             if self.shuffle:
-                self._shuffle_iterator()
+                self._shuffle()
+
             next_episode = next(self._iterator)
 
-        if self._prev_scene_id == next_episode.scene_id:
-            self._rep_count += 1
         if (
-            self.max_scene_repetition > 0
-            and self._rep_count >= self.max_scene_repetition - 1
+            self._prev_scene_id != next_episode.scene_id
+            and self._prev_scene_id is not None
         ):
-            self._shuffle_iterator()
             self._rep_count = 0
+            self._step_count = 0
 
         self._prev_scene_id = next_episode.scene_id
         return next_episode
 
-    def _shuffle_iterator(self) -> None:
+    def _switch_scene(self) -> None:
+        r"""Internal method to switch the scene
+        """
+        grouped_episodes = [
```

Comment:
It looks like this part of code doesn't respect `group_by_scene` as well as `shuffle` option.

Reply:
`groupby` creates a new group everytime the key changes and groups only get re-ordered if `shuffle` is true.

Reply:
So it'll respect both things :)

Reply:
I meant situation when `group_by_scene = False` is not considered.

Reply:
Yep, it'll also be fine in that situation. `groupby` doesn't do exactly what you would think, i.e.

```python
>> lst = ["a", "a", "b", "a"]
>> grouped = [list(g) for k, g in groupby(lst, lambda x: x)]
>> grouped
[['a', 'a'], ['b'], ['a']]
```